### PR TITLE
Fix/207 linux hot reloading

### LINKED ISSUE #207: Hot reloading does not work for subdirectories in Linux
#### Environment details

  - OS: Linux
  - Node.js version: 7.11.5
  - npm version: 3.10.10
  - @google-cloud/functions-emulator version: 1.0.0-beta.4

#### Steps to reproduce

  1. Create a functions folder where index.js is not at the root directory, but in a subdirectory such as 'lib/index.js' (this is a common pattern for users that write functions in TypeScript and then compile them to JS)
  2. Emulate the function
  3. Change the source code in lib/index.js
  4. Observe that the worker is not closed

After taking a look at the code, it seems from [this line](https://github.com/GoogleCloudPlatform/cloud-functions-emulator/blob/ca6f6d57ce53022b9b325bb2e6aa924c89944e52/src/supervisor/worker.js#L146) that `fs.watch` is used for hotloading. This has a known limitation that `recursive: true` does not work on Linux (see https://nodejs.org/api/fs.html#fs_caveats). In order to work around it, what do you think about checking `package.json` for the `main` field, which indicates the file that contains the functions, and then also watching the directory that file is in? But this would still lead to problems if the main file for functions calls helper functions in other subdirectories, so it doesn't solve the issue completely. Alternatively, the emulator can crawl the entire directory first so it knows all the subdirectories to watch, but that is a lot of work. So perhaps it's worth just printing out a warning messaging when Linux is detected.

### PR DESCRIPTION
**Pull request by sdbondi**
Fixes #207 (hopefully)

Using `chokidar` for file watching in the worker which fixes hot reloading on Linux.

### CODE REVIEW

File: src/supervisor/worker.js [hunk 1]
```diff
@@ -168,27 +170,28 @@ function main () {
...
-      }, (event, filename) => {
-        // Ignore node_modules
-        if (Array.isArray(message.watchIgnore)) {
-          for (let i = 0; i < message.watchIgnore.length; i++) {
-            if ((new RegExp(message.watchIgnore[i])).test(filename)) {
-              return;
-            }
-          }
-        }
+    if (localdir && message.watch) {
```
Comment (jmdobry) [inline src/supervisor/worker.js]: Why change this conditional? Reloading the function while in debug mode will break whatever debugging the user is doing if a file changes while they're debugging.

Comment (sdbondi) [reply src/supervisor/worker.js]: Thanks for reviewing! I think I wanted to test/debug the block of code, removed the conditional and then forgot to add it back (as you do :laughing:) - good catch.

**Comment by psigen** [comment]
Umm... does it normally take more than a month for CircleCI checks to complete? :sweat_smile:

**Comment by AustinGomez** [comment]
Any update here?

**Comment by hadyrashwan** [comment]
any idea when this fix is coming to the latest npm packge ??

**Comment by jmdobry** [comment]
Still a couple failing tests after the dependency upgrades, stay tuned...

**Pull request merged**